OSVR-Unity/Assets/OSVRUnity/src/VRViewer.cs

```diff
@@ -0,0 +1,147 @@
+/// </summary>
+
+using UnityEngine;
+using System.Collections;
+
+namespace OSVR
+{
+    namespace Unity
+    {
+        [RequireComponent(typeof(Camera))]
+        public class VRViewer : MonoBehaviour
+        {
+            #region Public Variables
+            public Camera Camera 
+            { 
+                get 
+                { 
+                    if(_camera == null)
+                    {
+                        _camera = GetComponent<Camera>();
+                    }
+                    return _camera; 
+                } 
+                set { _camera = value; } 
+            }
+            public DisplayController DisplayController { get { return _displayController; } set { _displayController = value; } }
+            [HideInInspector]
+            public Transform cachedTransform;
+            #endregion
+
+            #region Private Variables
+            private DisplayController _displayController;
+            private Camera _camera;
+            private bool disabledCamera = true;
+            #endregion
+
+            void Awake()
+            {
+                Init();
+            }
+
+            void Init()
+            {
+                //cache:
+                cachedTransform = transform;
+            }
+
+            void OnEnable()
+            {
+                StartCoroutine("EndOfFrame");
+            }
+
+            void OnDisable()
+            {
+                StopCoroutine("EndOfFrame");
+            }
+
+            //Updates the position and rotation of the head
+            public void UpdateViewerHeadPose(OSVR.ClientKit.Pose3 headPose)
+            {              
+                transform.localPosition = Math.ConvertPosition(headPose.translation);
+                transform.localRotation = Math.ConvertOrientation(headPose.rotation);
+            }
+
+            //Culling determines which objects are visible to the camera. OnPreCull is called just before this process.
+            void OnPreCull()
+            {
+                //update the client
+                _displayController.UpdateClient();
+
+                // Disable dummy camera during rendering
+                // Enable after frame ends
+                _camera.enabled = false;
+
+                //update the viewer's head pose
+                UpdateViewerHeadPose(_displayController.DisplayConfig.GetViewerPose(DisplayController.DEFAULT_VIEWER));
+
+                //render each eye camera (each surface)
+                //assumes one surface per eye
+                for (int i = 0; i < _displayController.EyeCount; i++)
+                {
+                    //update the eye pose
+                    VREye eye = _displayController.Eyes[i];
+                    eye.UpdateEyePose(_displayController.DisplayConfig.GetViewerEyePose(DisplayController.DEFAULT_VIEWER, (byte)i));
+
+                    //get the eye's surface
+                    VRSurface surface = eye.Surface;
+
```

Comment:
Note that I am not setting the camera.worldtoViewMatrix. The code below messes up rendering (invisible polys):
```c#
OSVR.ClientKit.Matrix44f viewMatrix = _displayController.DisplayConfig.GetViewerEyeViewMatrixf(
                        DisplayController.DEFAULT_VIEWER, (byte)i, OSVR.ClientKit.MatrixConventionsFlags.ColMajor);

                    surface.SetViewMatrix(Math.ConvertMatrix(viewMatrix));
```